src/core/quantum_circuit.jl

```diff
@@ -212,22 +223,99 @@ function simulate(circuit::QuantumCircuit)
     system = MultiBodySystem(circuit.qubit_count, 2)
     # initial state 
     ψ = fock(0, hilbert_space_size)
-    for step in circuit.pipeline
-        # U is the matrix corresponding the operations happening this step
-        #        U = Operator(Matrix{Complex}(1.0I, hilbert_space_size, hilbert_space_size))  
-        for gate in step
-            # if single qubit gate, get the embedded operator
-            # TODO: make sure embedding works for multi qubit system
-            S =
-                (length(gate.target) == 1) ?
-                get_embed_operator(gate.operator, gate.target[1], system) : gate
+    for step in circuit.pipeline 
+        for gate in step 
+            S = get_embed_operator(gate, system)
             ψ = S * ψ
         end
 
     end
     return ψ
 end
 
+function get_embed_operator(gate::Gate, system::MultiBodySystem)
+    if length(gate.target) == 1
+        return get_embed_operator(gate.operator, gate.target[1], system)
+    else
+        gate_to_operator =
+            Dict("cz"=>get_embed_controlled_gate_operator(sigma_z(), gate, system),
+            "cx"=>get_embed_controlled_gate_operator(sigma_x(), gate, system),
+            "iswap"=>get_swap(gate, system))
+        return gate_to_operator[gate.instruction_symbol]
+    end
+end
+
+function get_embed_controlled_gate_operator(controlled_operator::Operator, gate::Gate,
+    system::MultiBodySystem)
```

Comment:
This function only works with two-qubit gates. What if we have a three qubit gate such as Toffoli gate? We again need a more generic approach.